src/boat_simulator/boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,259 @@
+        Angles are bound with the convention (-180, 180]
+
+        Returns:
+            Scalar: The error between the given headings in radians, 0 radians at positive
+            y axis and increasing clockwise
+        """
+        desired_rad, current_rad = np.deg2rad(self.desired_heading), np.deg2rad(
+            self.current_heading
+        )
+
+        error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+
+        return error  # in radians
+
+    def compute_setpoint(self) -> Scalar:
+        """Computes the corresponding control error angle between current control angle and
+        target control angle. Uses Raye's implementation from:
+            https://github.com/UBCSailbot/raye-boat-controller/blob/master/python/tack_controller.py
+
+        Returns:
+            Scalar: Corresponding error angle between the
+            current and target control angle in degrees"""
+
+        heading_error = self.compute_error()  # heading_error in radians
+
+        rudder_setpoint = (self.kp * heading_error) / (1 + (self.cp * abs(heading_error)))
+
+        rudder_setpoint = min(
+            max(rudder_setpoint, np.deg2rad(self.max_angle_range[0])),
+            np.deg2rad(self.max_angle_range[1]),
+        )
+
+        rudder_setpoint_deg = np.rad2deg(rudder_setpoint)  # in degrees
+        self.running_error = rudder_setpoint_deg - self.current_control_ang  # in degrees
+        self.setpoint = rudder_setpoint_deg  # in degrees
+
+        return rudder_setpoint_deg  # in degrees
+
+    def reset_setpoint(self, new_desired_heading: Scalar, new_current_heading: Scalar) -> None:
+        """Resets a new desired heading angle, therefore recalculating the corresponding
+        setpoint and running error.
+
+        Args:
+            `new_desired_heading` (Scalar): New desired heading in degrees
+
+        """
+
+        self.desired_heading = new_desired_heading
+        self.current_heading = new_current_heading
+        self.compute_setpoint()
+
+    def change_desired_heading(self, changed_desired_heading) -> None:
+        """Changes desired heading to a new angle. Used for testing purposes
+
+        Args:
+            `changed_desired_heading` (Scalar): New desired heading in degrees
+
+        """
+        self.desired_heading = changed_desired_heading
+
+
+class SailController(ActuatorController):
+    """General Class for the Actuator Controller.
+
+    SailController Extends: ActuatorController
+    """
+
+    def __init__(
+        self,
+        target_angle: Scalar,
+        current_control_ang: Scalar,
+        time_step: Scalar,
+        kp: Scalar,
+        cp: Scalar,
+        control_speed: Scalar,
+        max_angle_range=SAIL_MAX_ANGLE_RANGE,
+    ):
+        """Initializes the class attributes.
+
+        Args:
+            `target_angle` (Scalar): Target angle for sail controller in degrees.
+             0 degrees (North) at the positive y axis and increasing clockwise.
```

Comment:
In particular, the target angle is for the **trim tab** on the sail, not the sail itself.